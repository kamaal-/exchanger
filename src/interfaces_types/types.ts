import Rates from "./Rates";
import Wallet from "./Wallet";
import Transaction from "./Transaction";

export type NullableRates = Rates | null;
export type NullableString = string | null;
export type NullableNumber = number | null;
export type NullableRate = Rate | null;
export type NullableWallets = Wallet[] | null;
export type NullableWallet = Wallet | null;
export type NullableBalance = Balance | null;

export type PairWallets = {
  fromWallet: Wallet;
  toWallet: Wallet;
};

export type Balance = {
  currency: string;
  balance: number;
  diff: number;
  label: string;
};

export type BalancePair = {
  [s: string]: Balance;
};

export type StoreState = {
  Rates: NullableRates;
  fromRate: Rate | null;
  toRate: Rate | null;
  error: NullableString;
  Wallets: WalletsState;
};

export type RatesState = {
  rates: NullableRates;
  fromRate: Rate | null;
  toRate: Rate | null;
  error: NullableString;
  base: NullableString;
};

export type WalletsState = {
  wallets: NullableWallets;
  fromWallet: NullableWallet;
  toWallet: NullableWallet;
  balances: BalancePair | null;
  leftWallet: NullableWallet;
  rightWallet: NullableWallet;
  transactions: Transaction[];
  value: string;
};

export type Action = {
  type: string;
  payload: any;
};

export type ConverterArgs = {
  fromType: string;
  toType: string;
  value: number;
  rates: Rates;
  base?: string;
  decimalPlaces?: number;
  fallback?: number;
};

export type Rate = {
  currency: string;
  rate: number;
};

export enum Size {
  Large = "Large",
  Medium = "Medium",
  Small = "Small"
}

export enum Align {
  Left,
  Right
}

export enum ButtonTypes {
  Main,
  Secondary,
  Disabled
}

export enum Theme {
  Dark,
  Light
}

export enum WalletType {
  From = "fromWallet",
  To = "toWallet"
}

export enum WalletSides {
  Left = "leftWallet",
  Right = "rightWallet"
}

// Props

export type InputProps = {
  wallet: Wallet;
  exchangeWallet: Wallet;
  theme: Theme;
  triggerConversion?: (
    input: string,
    fromWallet: Wallet,
    toWallet: Wallet
  ) => void;
  setFromWallet?: (w: Wallet) => void;
  setToWallet?: (w: Wallet) => void;
};

export type WalletProps = {
  wallet: Wallet;
  exchangeWallet: Wallet;
  theme: Theme;
  type: WalletType;
  balances: BalancePair;
  toWallet: Wallet;
  exchange: () => void;
};
