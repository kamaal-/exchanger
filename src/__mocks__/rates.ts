export const rates = {
    "EUR": {
        currency: 'EUR',
        rate: 0.9096697899
    },
    "SEK": {
        currency: 'SEK',
        rate: 9.9072136814
    },
    "USD": {
        currency: 'USD',
        rate: 1.0
    },
    "GBP": {
        currency: 'GBP',
        rate: 0.8110161012
    }
};

export const rate2 = {
    "EUR": {
        currency: 'EUR',
        rate: 1.9096697899
    },
    "SEK": {
        currency: 'SEK',
        rate: 9.9072136814
    },
    "USD": {
        currency: 'USD',
        rate: 1.0
    },
    "GBP": {
        currency: 'GBP',
        rate: 0.8110161012
    }
};