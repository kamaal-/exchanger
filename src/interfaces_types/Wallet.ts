import {Rate} from './types'
export default interface Wallet extends Rate {
    symbol: string;
    balance: number;
    id?:number;
    convert? :(value:number, exchangeWallet:Wallet) => number | null;
}