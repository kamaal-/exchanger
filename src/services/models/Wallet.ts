import IWallet from '../../interfaces_types/Wallet'
import {_convert, formateDecimalPlaces} from '../../utils'


export default class Wallet implements IWallet {

    currency: string = '';
    balance:number = 0;
    rate:number = 1;
    id?:number;
    symbol: string = '';
    constructor(args?:IWallet, id?:number){
        const {rate = 1, currency = '', balance = 0, symbol = ''} = args ? args : {};
        this.balance = balance
        this.rate = rate
        this.currency = currency
        this.id = id;
        this.symbol = symbol ;
    }

    convert(value:number | string, toWallet: IWallet) : number | null{
        const converted = _convert(value, this, toWallet, 'USD');
        return converted ? formateDecimalPlaces(converted) : null;
    }
}