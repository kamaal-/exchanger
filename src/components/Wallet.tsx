import React, { useState } from "react";
import { useSpring } from "react-spring";
import {
  WalletWrapper,
  WalletTitle,
  CurrencyLogo,
  ExhangeRate,
  WalletBalance,
  Button,
  WalletRow
} from "../theme/index";
import { Size, Align, WalletProps } from "../interfaces_types/types";
import { formateDecimalPlaces } from "../utils";
import Input from "./Input";
import Balance from "./Balance";

const Wallet = ({
  wallet,
  exchangeWallet,
  theme,
  type,
  exchange,
  toWallet
}: WalletProps) => {
  const styleProps = useSpring({
    from: { opacity: "0" },
    to: { opacity: "1" }
  });
  const exchangeRate = wallet.convert
    ? wallet.convert(1, exchangeWallet)
    : null;
  const [buttonIsActive, setButtonState] = useState<Boolean>(false);
  const isToWallet = wallet.currency === toWallet.currency;
  return (
    <WalletWrapper style={styleProps}>
      <WalletRow>
        <CurrencyLogo data-testid={`wallet-logo`} size={Size.Large} currency={wallet.currency}>
          {wallet.symbol}
        </CurrencyLogo>
        <WalletTitle data-testid={`wallet-currency`}>{wallet.currency}</WalletTitle>
        <ExhangeRate data-testid={`wallet-exchange`} size={Size.Large} align={Align.Right}>
          {exchangeRate
            ? `1 ${wallet.currency} = ${exchangeRate} ${exchangeWallet.currency}`
            : ""}
        </ExhangeRate>
      </WalletRow>
      <WalletRow>
        <WalletBalance data-testid={`wallet-balance`}>
          You have {formateDecimalPlaces(wallet.balance)} {wallet.currency}
        </WalletBalance>
      </WalletRow>
      <Input
        setButtonState={setButtonState}
        exchangeWallet={exchangeWallet}
        wallet={wallet}
        theme={theme}
      />
      <WalletRow style={{ alignItems: "flex-end" }}>
        <Balance type={type} />
        <Button data-testid={`wallet-button`}
          onClick={() => {
            if (buttonIsActive) {
              // do the math
              exchange();
            }
          }}
          active={buttonIsActive && !isToWallet}
          style={{ width: "100px", marginLeft: "auto" }}
        >
          EXCHANGE
        </Button>
      </WalletRow>
    </WalletWrapper>
  );
};

export default Wallet;
