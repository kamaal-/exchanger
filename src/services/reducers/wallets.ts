import { WalletsState, Action } from "../../interfaces_types/types";
import { WALLETS, CONVERT, APP } from "../action_types/index";

export const initialState: WalletsState = {
  wallets: null,
  fromWallet: null,
  toWallet: null,
  balances: null,
  leftWallet: null,
  rightWallet: null,
  transactions: [],
  value: ""
};

export default (state = initialState, { type, payload }: Action) => {
  switch (type) {
    case WALLETS.ADD:
      return { ...state, wallets: payload };
    case WALLETS.REMOVE:
      return { ...state, wallets: payload };
    case WALLETS.FROM:
      return { ...state, fromWallet: payload };
    case WALLETS.TO:
      return { ...state, toWallet: payload };
    case WALLETS.SWITCH:
      return {
        ...state,
        toWallet: state.fromWallet,
        fromWallet: state.toWallet
      };
    case CONVERT.DO:
      return { ...state };
    case WALLETS.LEFT:
      return { ...state, leftWallet: payload };
    case WALLETS.RIGHT:
      return { ...state, rightWallet: payload };
    case WALLETS.RESET:
      return {
        ...state,
        balances: payload.balances,
        fromWallet: payload.fromWallet,
        toWallet: payload.toWallet
      };
    case CONVERT.RUN:
      return { ...state, balances: payload };
    case WALLETS.ADD_TRANSACTION:
      return { ...state, transactions: payload };
    case WALLETS.SET_VALUE:
      return { ...state, value: payload };
    case APP.SUCCESS_POLLING:
    case APP.START_POLLING:
      return { ...state };
    default:
      return initialState;
  }
};
