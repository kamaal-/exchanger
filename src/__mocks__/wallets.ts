import IWallet from "../interfaces_types/Wallet";
import { rates } from "./rates";
import Wallet from "../services/models/Wallet";
import { symbol } from "prop-types";

export const wallets: IWallet[] = [
  {
    currency: rates["EUR"].currency,
    rate: rates["EUR"].rate,
    balance: 10.23,
    symbol: "€"
  },
  {
    currency: rates["USD"].currency,
    rate: rates["USD"].rate,
    balance: 150.45,
    symbol: "$"
  },
  {
    currency: rates["GBP"].currency,
    rate: rates["GBP"].rate,
    balance: 320.4,
    symbol: "l"
  }
];

export const balances = {
  fromWallet: {
    currency: "EUR",
    balance: 10.23,
    diff: 0,
    label: "Balance"
  },
  toWallet: {
    currency: "USD",
    balance: 150.45,
    diff: 0,
    label: "Balance"
  }
};

export const balances_ok = {
  fromWallet: {
    currency: "EUR",
    balance: 10.23,
    diff: -3.23,
    label: "Balance"
  },
  toWallet: {
    currency: "USD",
    balance: 150.45,
    diff: 4.56,
    label: "Balance"
  }
};

export const balances_not_ok = {
  fromWallet: {
    currency: "EUR",
    balance: 10.23,
    diff: -10.24,
    label: "Balance"
  },
  toWallet: {
    currency: "USD",
    balance: 150.45,
    diff: 4.56,
    label: "Balance"
  }
};


export const transactions = [ 
  { 
     "detail":"You have transfered 34 kr to RM.",
     "wallets":{ 
        "from":{ 
           "currency":"SEK",
           "symbol":"kr"
        },
        "to":{ 
           "currency":"MYR",
           "symbol":"RM"
        }
     },
     "from":{ 
        "before":"100 kr",
        "diff":"34 kr",
        "after":"66 kr"
     },
     "to":{ 
        "before":"200 RM",
        "diff":"14.47 RM",
        "after":"214.47 RM"
     },
     "exchange":"1 kr = 0.42 RM",
     "time":1570979215945
  },
  { 
     "detail":"You have transfered 2 € to £.",
     "wallets":{ 
        "from":{ 
           "currency":"EUR",
           "symbol":"€"
        },
        "to":{ 
           "currency":"GBP",
           "symbol":"£"
        }
     },
     "from":{ 
        "before":"95 €",
        "diff":"2 €",
        "after":"93 €"
     },
     "to":{ 
        "before":"104.38 £",
        "diff":"1.75 £",
        "after":"106.13 £"
     },
     "exchange":"1 € = 0.87 £",
     "time":1570979187093
  },
  { 
     "detail":"You have transfered 5 € to £.",
     "wallets":{ 
        "from":{ 
           "currency":"EUR",
           "symbol":"€"
        },
        "to":{ 
           "currency":"GBP",
           "symbol":"£"
        }
     },
     "from":{ 
        "before":"100 €",
        "diff":"5 €",
        "after":"95 €"
     },
     "to":{ 
        "before":"100 £",
        "diff":"4.38 £",
        "after":"104.38 £"
     },
     "exchange":"1 € = 0.87 £",
     "time":1570973116868
  }
];