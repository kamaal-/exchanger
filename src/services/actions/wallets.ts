import { WALLETS, CONVERT } from "../action_types/";
import { ThunkDispatch } from "redux-thunk";
import { saveInLocalStorage } from "../../utils/";
import Wallet from "../../interfaces_types/Wallet";
import { StoreState, Balance, BalancePair } from "../../interfaces_types/types";
import Transaction from "../models/Transaction";
import ITransaction from "../../interfaces_types/Transaction";

export const setFromWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  let { balances } = getState().Wallets;
  balances = balances ? { ...balances } : {};
  await saveInLocalStorage("fromWallet", wallet);
  dispatch({ type: WALLETS.FROM, payload: wallet });
  balances["fromWallet"] = {
    currency: wallet.currency,
    balance: wallet.balance,
    diff: 0,
    label: "Balance"
  };
  dispatch({ type: CONVERT.RUN, payload: balances });
};

export const setLeftWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  await saveInLocalStorage("leftWallet", wallet);
  dispatch({ type: WALLETS.LEFT, payload: wallet });
};

export const setToWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  let { balances } = getState().Wallets;
  balances = balances ? { ...balances } : {};
  await saveInLocalStorage("toWallet", wallet);
  dispatch({ type: WALLETS.TO, payload: wallet });
  balances["toWallet"] = {
    currency: wallet.currency,
    balance: wallet.balance,
    diff: 0,
    label: "After"
  };
  dispatch({ type: CONVERT.RUN, payload: balances });
};

export const setRightWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  await saveInLocalStorage("rightWallet", wallet);
  dispatch({ type: WALLETS.RIGHT, payload: wallet });
};

export const switcWallets = () => (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  let { fromWallet, toWallet } = getState().Wallets;
  if (fromWallet && toWallet) {
    dispatch(setToWallet(fromWallet));
    dispatch(setFromWallet(toWallet));
    dispatch(triggerConversion("0", toWallet, fromWallet));
  }
};

export const resetWallets = (fromWallet: Wallet, toWallet: Wallet) => (
  dispatch: ThunkDispatch<{}, {}, any>
) => {
  if (fromWallet && toWallet) {
    const balances = {
      fromWallet: {
        currency: fromWallet.currency,
        balance: fromWallet.balance,
        diff: 0,
        label: "Balance"
      },
      toWallet: {
        currency: toWallet.currency,
        balance: toWallet.balance,
        diff: 0,
        label: "Balance"
      }
    };

    dispatch({
      type: WALLETS.RESET,
      payload: { balances, fromWallet, toWallet }
    });
    dispatch(setValue(""));
  }
};

export const saveTransaction = (
  transaction: ITransaction,
  key = "transactions"
) => (dispatch: ThunkDispatch<{}, {}, any>) => {
  if (transaction && transaction instanceof Transaction) {
    let transactionsList: ITransaction[] = [];
    try {
      const oldTransactions = localStorage.getItem(key);
      if (oldTransactions) {
        transactionsList = JSON.parse(oldTransactions);
      }
      transactionsList = transactionsList.map(t => {
        const obj = new Transaction();
        obj.convenientConsttructor(t);
        return obj;
      });
      transactionsList.unshift(transaction);
      saveInLocalStorage(key, transactionsList);
      dispatch({ type: WALLETS.ADD_TRANSACTION, payload: transactionsList });
    } catch (error) {
      transactionsList.unshift(transaction);
      saveInLocalStorage(key, transactionsList);
      dispatch({ type: WALLETS.ADD_TRANSACTION, payload: transactionsList });
    }
  }
};

export const triggerConversion = (
  input: string,
  fromWallet: Wallet,
  toWallet: Wallet
) => (dispatch: ThunkDispatch<{}, {}, any>) => {
  if (input && fromWallet && toWallet) {
    dispatch({ type: CONVERT.DO, payload: { input, fromWallet, toWallet } });
  }
};

export const getTransaction = (
  fromWallet: Wallet,
  toWallet: Wallet,
  balances: BalancePair
) => new Transaction(fromWallet, toWallet, balances);

export const exchange = (cb = () => {}) => (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  const state: StoreState = getState();
  const { fromWallet, toWallet, balances, wallets } = state.Wallets;

  // detuct from wallet
  if (fromWallet && toWallet && balances) {
    const fromBalance: Balance = balances["fromWallet"];
    const toBalance = balances["toWallet"];
    // only pass the exchange if their any values and value can deduct from purse.
    if (
      fromBalance.diff &&
      toBalance.diff &&
      Math.abs(fromBalance.diff) <= fromWallet.balance
    ) {
      // keep the transaction before update wallets
      const transaction = getTransaction(fromWallet, toWallet, balances);

      fromWallet.balance = fromBalance.balance;
      toWallet.balance = toBalance.balance;
      const newWallets = (wallets || [])
        .map(w => (w.currency === fromWallet.currency ? fromWallet : w))
        .map(w => (w.currency === toWallet.currency ? toWallet : w));
      dispatch(updateWallets(newWallets));
      dispatch(setFromWallet(fromWallet));
      dispatch(setToWallet(toWallet));
      const newBalances = {
        fromWallet: {
          currency: fromWallet.currency,
          balance: fromWallet.balance,
          diff: 0,
          label: "Balance"
        },
        toWallet: {
          currency: toWallet.currency,
          balance: toWallet.balance,
          diff: 0,
          label: "Balance"
        }
      };
      dispatch({
        type: WALLETS.RESET,
        payload: { balances: newBalances, fromWallet, toWallet }
      });
      dispatch(saveTransaction(transaction));
      dispatch(resetWallets(fromWallet, toWallet));
      dispatch(setValue(""));
    }
  }
};

export const updateWallets = (wallets: Wallet[]) => async (
  dispatch: ThunkDispatch<{}, {}, any>
) => {
  if (wallets && wallets.length >= 0) {
    await saveInLocalStorage("wallets", wallets);
    dispatch({ type: WALLETS.ADD, payload: wallets });
  }
};

export const setValue = (value: string) => async (
  dispatch: ThunkDispatch<{}, {}, any>
) => {
  dispatch({ type: WALLETS.SET_VALUE, payload: value });
};
