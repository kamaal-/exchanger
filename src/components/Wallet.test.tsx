import React from 'react'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import Wallet from './Wallet'
import { Theme} from '../interfaces_types/types'
import { render } from '@testing-library/react'
import {balances, wallets} from '../__mocks__/wallets'

const wallet = wallets[0];

describe('Input component', () => {
    function renderWithRedux(
        ui,
        { _, store = createStore(() => ({Wallets:{balances}})) } = {}
      ) {
        return {
          ...render(<Provider store={store}>{ui}</Provider>),
          // adding `store` to the returned utilities to allow us
          // to reference it in our tests (just try to avoid using
          // this to test implementation details).
          store,
        }
    }

    const setup = () => {
        return renderWithRedux(<Wallet wallet={wallet} exchangeWallet={wallets[1]} toWallet={wallets[1]} theme={Theme.Light} balances={balances} exchange={() => {}} />)
    }

    it('should match snapshot', () => {
        const { asFragment } = setup();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render with logo', () => {
        const {getByTestId} = setup()
        expect(getByTestId(`wallet-logo`).textContent).toBe('â‚¬')
    });

    it('should render with currency', () => {
        const {getByTestId} = setup()
        expect(getByTestId(`wallet-currency`).textContent).toBe('EUR')
    });

    it('should render with exchange', () => {
        const {getByTestId} = setup()
        expect(getByTestId(`wallet-exchange`).textContent).toBe('')
    });

    it('should render with balance', () => {
        const {getByTestId} = setup()
        expect(getByTestId(`wallet-balance`).textContent).toBe('You have 10.23 EUR')
    });

    it('should render with button', () => {
        const {getByTestId} = setup()
        expect(getByTestId(`wallet-button`).textContent).toBe('EXCHANGE')
    });
});