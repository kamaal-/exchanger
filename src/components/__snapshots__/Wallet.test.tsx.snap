// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa cRcnPO"
    style="opacity: 0;"
  >
    <div
      class="sc-dnqmqq dOtUEt"
    >
      <div
        class="sc-bZQynM dCbpqh"
        data-testid="wallet-logo"
      >
        €
      </div>
      <h2
        class="sc-bwzfXH laokzr"
        data-testid="wallet-currency"
      >
        EUR
      </h2>
      <div
        class="sc-EHOje xMYza"
        data-testid="wallet-exchange"
      />
    </div>
    <div
      class="sc-dnqmqq dOtUEt"
    >
      <div
        class="sc-htpNat cSeCBV"
        data-testid="wallet-balance"
      >
        You have 10.23 EUR
      </div>
    </div>
    <div
      class="sc-dnqmqq dOtUEt"
      data-testid="EUR-input"
    >
      <input
        class="sc-bxivhb imKKaO"
        data-testid="revolut-input"
        placeholder="0"
        value=""
      />
      <div
        class="sc-ifAKCX gsjozh"
      >
        €
      </div>
    </div>
    <div
      class="sc-dnqmqq dOtUEt"
      style="align-items: flex-end;"
    >
      <div
        class="sc-gzVnrw kNwIxh"
        data-testid="wallet-button"
        style="width: 100px; margin-left: auto;"
      >
        EXCHANGE
      </div>
    </div>
  </div>
</DocumentFragment>
`;
